iml-services/iml-devices/src/db.rs

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+//! # DB
+//!
+//! This module is a host side plugin to persist devices to the db.
+//!
+//! It populates two tables, `device` and `device_host`.
+//!
+//! A `device` is a generic device that may be present on multiple hosts.
+//!
+//! A `device_host` is the instance of that device on a host.
+//!
+//! `device` to `device_host` is a 1:M relationship.
+//!
+//! Whenever a message is received, `Device` and `DeviceHost` records are loaded from the
+//! database. The incoming tree is converted to a db compatible format, and is merged with the
+//! existing database records. The complete tree is then checked to see if any virtual devices
+//! need to be added, removed or updated based on their parents. A changelog is kept and the database is
+//! then patched up with any changes.
+//!
+//! The choice to persist to the database is because:
+//!
+//! 1. The need to have a "balanced" set of Lustre targets, and for this balance to be stable across ticks of
+//!    device discovery.
+//! 2. The need for devices to not be forgotten if they are in use as Lustre targets. Instead this should be an alert.
+//!
+
+use crate::{
+    change::{self, Change},
+    error::ImlDevicesError,
+};
+use iml_postgres::Transaction;
+use iml_wire_types::{
+    db::{Device, DeviceHost, DeviceId, DeviceIds, DeviceType, MountPath, Paths, Size},
+    Fqdn,
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    path::PathBuf,
+};
+
+pub type FlatDevices = BTreeMap<DeviceId, FlatDevice>;
+
+pub type Devices = BTreeMap<DeviceId, Device>;
+
+pub type DeviceHosts = BTreeMap<DeviceId, DeviceHost>;
+
+/// A device (Block or Virtual).
+/// These should be unique per cluster
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct FlatDevice {
+    pub id: DeviceId,
+    pub size: u64,
+    pub device_type: DeviceType,
+    pub parents: BTreeSet<DeviceId>,
+    // pub usable_for_lustre: bool,
```

Comment:
Why is it commented out?

Reply:
This needs to be implemented in `flat_devices.rs`.